stream-parent/stream-client/src/main/java/org/interledger/stream/sender/DefaultBackoffController.java

```diff
@@ -0,0 +1,109 @@
+import org.interledger.core.InterledgerPreparePacket;
+import org.interledger.core.InterledgerRejectPacket;
+import org.interledger.core.InterledgerResponsePacket;
+import org.interledger.link.Link;
+import org.interledger.link.exceptions.LinkRetriesExceededException;
+
+import io.github.resilience4j.retry.IntervalFunction;
+import io.github.resilience4j.retry.Retry;
+import io.github.resilience4j.retry.RetryConfig;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+public class DefaultBackoffController implements BackoffController {
+
+  private final Retry retry;
+
+  private final int maxAttempts;
+
+  public DefaultBackoffController() {
+    this(Duration.ofMillis(500), 2.0, 5);
+  }
+
+  public DefaultBackoffController(final Duration duration, final double multiplier,
+                                  final int maxAttempts) {
+    this.maxAttempts = maxAttempts;
+    // use random backoff to avoid multiple backed off packets from hitting concurrently
+    IntervalFunction interval = IntervalFunction.ofExponentialRandomBackoff(duration, multiplier);
+    RetryConfig config = constructConfig(maxAttempts, interval);
+    retry = Retry.of("DefaultBackoffControllerRetry", config);
+  }
+
+  protected RetryConfig constructConfig(int maxAttempts, IntervalFunction interval) {
+    return RetryConfig.custom()
+        .maxAttempts(maxAttempts)
+        .intervalFunction(interval)
+        .retryOnResult(responsePacket -> {
+          AtomicBoolean retry = new AtomicBoolean(false);
+          ((InterledgerResponsePacket) responsePacket).handle(
+            (fulfill) -> retry.set(false),
+            (reject) -> retry.set(rejectIsRetryable(reject))
+          );
+          return retry.get();
+        })
+        .build();
+  }
+
+  protected boolean rejectIsRetryable(InterledgerRejectPacket reject) {
+    // if we have a temporary error but it's not due to liquidity, then we're safe to retry
+    return reject.getCode().getErrorFamily() == InterledgerErrorCode.ErrorFamily.TEMPORARY &&
+      reject.getCode() != T04_INSUFFICIENT_LIQUIDITY;
+  }
+
+  public static class LinkSendRejectTrackingWrapper {
+
+    private final AtomicInteger rejects = new AtomicInteger(0);
+
+    private final Link link;
+
+    public LinkSendRejectTrackingWrapper(Link link) {
+      this.link = link;
+    }
+
+    public InterledgerResponsePacket sendWrapped(InterledgerPreparePacket preparePacket) {
+      InterledgerResponsePacket response = link.sendPacket(preparePacket);
+      response.handle(
+        (fulfill) -> {},
+        (reject) -> rejects.incrementAndGet()
+      );
+      return response;
+    }
+
+    public int rejectsReceived() {
+      return rejects.get();
+    }
+  }
+
+  @Override
+  public InterledgerResponsePacket sendWithBackoff(final Link link, final InterledgerPreparePacket preparePacket,
```

Comment:
I think that the preparePacket.expiresAt would need to be updated/extended before every retry attempt. When SimpleStreamSender creates the preparePacket it sets the expiresAt to 30s in the future. If you configured back such that it retries after 1s, 2s, 4s, 8s, 16s then the logic attempts retry a preparePacket that is already expired.

Reply:
Good catch; totally spaced that behavior. Thanks!

Reply:
I was expecting that if a packet fails, we would consider it a reject, and the "retry" would just be a new packet. Is the design here to do something different?